### ISSUE TITLE
OAuth Server and Default Callback Support

### ISSUE THREAD

**Issue by michaelrsweet** [OpenPrinting/cups#246]
This bug is a placeholder for adding OAuth server support to cupsd and figuring out the right default callbacks on different platforms.

**Comment by michaelrsweet** [comment]
Related to issue #100 which adds the basic client callback support in libcups.

**Comment by ShivanshCharak** [comment]
Okay so by authorization you mean when we open the localhost:631 it asks for username and password of the machine there we have to add authorization?

**Comment by michaelrsweet** [comment]
@ShivanshCharak No, this is to integrate support for using OAuth/OpenID with CUPS - on the client side to open a browser window to authorize access, and on the server (cupsd) side to "introspect" the access token to get the authorized user name.

In short, an alternative to the usual username + password stuff.

**Comment by yarons** [comment]
It's easier to manage from security perspective by allowing having a single identity per employee and auditing (SAML is a good example, think about controlling the paper type and color/b&w centrally in that case).
OpenID Connect is usually for public services authentication such as Google, Facebook, etc., I don't think it's mandatory for such project.

**Comment by ShivanshCharak** [comment]
@michaelrsweet okay i wanna work on this issue

**Comment by ShivanshCharak** [comment]
@michaelrsweet okay so you mean by using openid we have to do Authorization by poping up the window there we have to give our detail like email or password and by using tge system will give us access

**Comment by michaelrsweet** [comment]
@ShivanshCharak I appreciate the enthusiasm!

We are finalizing some OAuth things over in the Printer Working Group now, and I hope to post an initial document for discussion this week (I'll add a link here).

Aside from the logistics of *what* bits of OAuth/OpenID to support, there are also a bunch of security things to think about in implementation to prevent arbitrary programs from collecting access tokens.

**Comment by michaelrsweet** [comment]
OK, so I've been noodling some things for how to implement this:

- [x] Add an Authorization header store per FQDN (like the "ssl" directory, but for OAuth) with a simple API for getting/setting/clearing tokens
- [x] Also support a `CUPS_HTTP_AUTHORIZATION` environment variable to specify the Authorization: header to use for non-interactive usage
- [x] Add a command (`cupsoauth`?) to manage the encrypted store (get/set/clear)
- [x] Add cupsd support for `OAuthServer URI` directive and `AuthType Bearer` value to cupsd.conf, with requirement for OpenID/RFC 8414 metadata and JWT usage so that tokens can be introspected locally

For GUI applications, the CPDB UI will need to handle bringing up the OAuth authorization page and collecting the bearer token.

**Comment by michaelrsweet** [comment]
A lot of this is already implemented. What remains is the server side, so re-titling...

**Comment by michaelrsweet** [comment]
Some details as I'm getting ready to add this:

- Add MS Entrada ID client_id information to CUPS source code, which becomes the default client_id if none is stored for the current auth server and the auth server URL matches the MS "https://login.microsoftonline.com/TENANT/v2.0" pattern.
- New cups-files.conf directives:
  - `OAuthGroup NAME FILENAME`: File lists JWT "email" or "sub" values that should map to "@" groups, one per line
  - `OAuthAuthorizationServer https://hostname/PATH`: URL for authorization server
  - `OAuthAuthorizationScopes SCOPE [... SCOPE]`: Lists the scopes to ask for, defaults to all of the supported scopes if not specified
- New cupsd.conf `[Default]AuthType Bearer` directive to enable OAuth
- Map OAuth "name" to requesting-user-name/job-originating-user-name attributes
- Map OAuth "email" to requesting-user-uri/job-originating-user-uri attributes
- Update "@GROUP" auth code to use OAuth lookups for OAuth accounts

**Comment by michaelrsweet** [comment]
Also:

- Pass current Bearer token in environment (`HTTP_AUTHORIZATION`) to CGI programs, so they can pass the value through when sending requests.

**Comment by michaelrsweet** [comment]
Created a sub-issue to add proper support for OAuth authorization through the web interface (requires redirection - localhost will be easier than arbitrary servers...)

**Comment by michaelrsweet** [comment]
Another sub-issue for shared queues.

OK, basic support is pushed now, requires testing (obviously) and I'll work on some documentation as well. The basics are as follows:

Configuration
--------------

Add `OAuthServer` to "cups-files.conf" with the URL of your OAuth authorization server. For example, to use Microsoft's Entrada ID:

    OAuthServer https://login.microsoftonline.com/common/v2.0

Create a "SYSTEM.txt" file in "/etc/cups" and list any emails you want to have admin access. Then add an `OAuthGroup` line to "cups-files.conf":

    OAuthGroup SYSTEM /etc/cups/SYSTEM.txt

**Comment by michaelrsweet** [comment]
Closing this one, remaining tasks are separate issues.
